File: PAT_Basic_Level_by_cpp/1116 多二了一点/1116 多二了一点.cpp
// 1116 多二了一点
//
// 实现原理：
// 若一个正整数有 2n 个数位，后 n 个数位组成的数恰好比前 n 个数位组成的数多 2，
// 则称这个数字"多二了一点"。给定一个正整数 N（可达 10^1000），判断它是否多二了一点。
//
// 关键步骤：
// 1. 若 N 的位数是奇数，输出 "Error: X digit(s)"，X 为位数。
// 2. 否则将 N 从中间分成前一半 Y 和后一半 X。
// 3. 用大数减法计算 X - Y（题目保证多二了一点时 Y 的个位数不大于 7，无需借位）。
// 4. 若差为 2 输出 "Yes: X - Y = 2"，否则输出 "No: X - Y != 2"。
//
// 复杂度分析：
// 时间复杂度：O(L)，L 为 N 的位数。
// 空间复杂度：O(L)，存储字符串。
#include <iostream>
#include <string>

using namespace std;

int main() {
    string s;
    if (!(cin >> s)) return 0;

    int len = (int)s.size();
    if (len % 2 != 0) {                       // 奇数位
        cout << "Error: " << len << " digit(s)" << endl;
        return 0;
    }

    string Y = s.substr(0, len / 2);          // 前一半
    string X = s.substr(len / 2);             // 后一半

    // 只有 X >= Y 时才进行无符号大数减法；否则差值为负，不可能等于 2。
    bool xAtLeastY = (X >= Y); // 两个等长数字字符串可直接按字典序比较
    string diff;
    if (xAtLeastY) {
        int borrow = 0;
        for (int i = (int)X.size() - 1; i >= 0; --i) {
            int d = (X[i] - '0') - (Y[i] - '0') - borrow;
            if (d < 0) { d += 10; borrow = 1; } else borrow = 0;
            diff = char('0' + d) + diff;
        }
    }
    // 去掉前导零
    size_t pos = diff.find_first_not_of('0');
    string dstr = (pos == string::npos) ? "0" : diff.substr(pos);

    if (xAtLeastY && dstr == "2") cout << "Yes: " << X << " - " << Y << " = 2" << endl;
    else cout << "No: " << X << " - " << Y << " != 2" << endl;
    return 0;
}

File: PAT_Basic_Level_by_perl/1116 多二了一点/1116 多二了一点.pl
use strict;
use warnings;

# =====================================================================
# 题目：1116 多二了一点
# 难度：简单
# 知识点：字符串, 大数比较
# =====================================================================
#
# 【题目描述】
#   若正整数有偶数个数位，后一半数位组成的数恰好比前一半数位
#   组成的数多 2，则称"多二了一点"。判断输入整数的情况并输出。
#
# 【实现原理】
#   1. 若位数为奇数，输出 Error: X digit(s)。
#   2. 否则取前后两半数字，判断后一半减前一半是否等于 2。
#   3. 按 Yes/No 对应格式输出。
#
# 【输入样例】
#   233235
# 【输出样例】
#   Yes: 235 - 233 = 2
# =====================================================================

# 读取输入字符串（可能很长，不能转成数值）
my $s = <STDIN>;
chomp $s;

sub normalize {
    my ($v) = @_;
    $v =~ s/^0+//;
    return $v eq '' ? '0' : $v;
}

sub cmp_abs {
    my ($a, $b) = @_;
    return length($a) <=> length($b) || $a cmp $b;
}

sub subtract_abs {
    my ($a, $b) = @_;       # 要求 a >= b
    my @a = split //, $a;
    my @b = split //, $b;
    my @out;
    my ($i, $j, $borrow) = ($#a, $#b, 0);
    while ($i >= 0) {
        my $x = $a[$i] - $borrow;
        my $y = $j >= 0 ? $b[$j] : 0;
        if ($x < $y) {
            $x += 10;
            $borrow = 1;
        } else {
            $borrow = 0;
        }
        unshift @out, $x - $y;
        $i--;
        $j--;
    }
    return normalize(join('', @out));
}

if (length($s) % 2) {
    # 奇数个数位
    print "Error: ", length($s), " digit(s)\n";
}
else {
    my $h = int(length($s) / 2);
    my $a = normalize(substr($s, 0, $h));  # 前一半
    my $b = normalize(substr($s, $h));     # 后一半

    my $cmp = cmp_abs($b, $a);
    my ($diff_sign, $diff_abs);
    if ($cmp >= 0) {
        $diff_sign = '';
        $diff_abs = subtract_abs($b, $a);
    } else {
        $diff_sign = '-';
        $diff_abs = subtract_abs($a, $b);
    }
    my $diff = $diff_sign . $diff_abs;

    if ($diff eq '2') {
        print "Yes: $b - $a = 2\n";
    }
    else {
        print "No: $b - $a != 2\n";
    }
}
